output: add skybox support

lahm86 opened the pull request:
Resolves #94.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This adds support for skybox rendering similar to TR2. A default injection is provided for Lost Valley and Colossuem; custom level builders can use object slot 184, and in that case the config option is effectively overwritten so the builder has control over the level design.

The refactoring in `output.c` and `s_output.c` are courtesy of @rr- (thanks again) and fix a persistent clipping issue at the top of the screen we found in initial testing. An additional change in `Output_CalcRoomVertices` avoids room geometry disappearing beyond fog end, which can be quite jarring when fog drawing changes as Lara moves - this way the geometry is always visible in front of the sky, but draw distance still honoured.

The additional `.bin` change for Lost Valley fixes some missing textures that we had previously overlooked - these were at the top of the valley, and only really noticeable with the sky in place.

lahm86 committed:
output: add skybox support

This adds support for skybox rendering similar to TR2. A default
injection is provided for Lost Valley and Colossuem; custom level
builders can use object slot 184.

Resolves #94.

rr- commented on src/game/output.c:
```diff
@@ -320,7 +346,6 @@ static const int16_t *Output_CalcRoomVertices(const int16_t *obj_ptr)
             int32_t depth = zv_int >> W2V_SHIFT;
             if (depth > Output_GetDrawDistMax()) {
                 m_VBuf[i].g = 0x1FFF;
-                clip_flags |= 16;
```
1. Can we restore the original code and fence it with a condition `if (!g_Config.enable_skybox)`? This should add clarity to the code, especially when cross-referencing it with TR2+ outside flag.
2. Do we know how this is going to affect the rendering performance – do we now draw the entire level geometry, or do we still clip unseen rooms?

lahm86 replied:
1. Done - I've referenced the object loaded state though in case a custom level has the object, but the player disables the setting, as that might come across as a level issue. LMK if this looks OK. My idea was the config option only controls the injection.
2. Performance seems good to me. Yes, `Room_SetBounds` will cull rooms that are too far away anyway, so the above only applies to rooms in the "immediate" vicinity.
https://github.com/lahm86/TR1X/blob/develop/src/game/room_draw.c#L101

rr- commented on src/game/room_draw.c:
```diff
@@ -248,6 +250,24 @@ static void Room_PrepareToDraw(int16_t room_num)
+
+    Output_SetupAboveWater(g_Camera.underwater);
+    Matrix_Push();
+    g_MatrixPtr->_03 = g_MatrixPtr->_13 = g_MatrixPtr->_23 = 0;
```
Let's avoid chaining assignments in a single statement.

lahm86 committed:
Address comments

rr- reviewed (commented):
I'm leaning towards extending the skybox feature to apply to custom levels as well rather than injections alone. This will make the option more versatile and enable players who choose so, to experience true TR1 vibe in custom levels as well. Since the game settings are bound to each game directory, rather than being shared across all games, it shouldn't cause unnecessary surprises either.

Let's hear what others think.
@walkawayy @Richard-L @aredfan

Richard-L commented:
I never play TRLEs so don't have the most useful opinion I'm afraid. I would only extend it to them if the cost in code hygiene and maintainability is well manageable for you guys. I do like the general sentiment though.

Bigger question for me, perhaps worthy of a discord thread: what skyboxes would be used?

rr- commented:
IMO the best choices are skyboxes that have a 1996 aesthetic, and preferably coming from the original games. Of all the classic skyboxes available, I believe Lahm has  made the perfect choice with his selection as it looks perfect in both levels. Originating from TR2 Gold, it is also particularly appealing as it can be freely redistributed.

For the reference, here are WADs with all the classic skyboxes available: https://www.trsearch.org/item/5529
The only other option I'd consider is the Meteorite Cavern skybox, with a starry night sky and the Moon.

Richard-L commented:
Sounds good. I appreciate your desire to stick to the OG classics look.

I think some of them may _just_ benefit from certain Photoshop love perhaps. Not as much as what we did with the loading screens, but in that vein.

lahm86 committed:
Add vertex shade injections

This is used in Obelisk to effectively add enforced fog to the windows
into City of Khamoon.

lahm86 committed:
Create separate sky injections

lahm86 committed:
Make config option apply to all scenarios

This will allow players to disable the skybox in custom levels.

lahm86 commented:
Latest commits:
- Three separate skyboxes for the relevant levels
- The gaps in Obelisk into City will now have perma-fog in effect (bundled with texture fixes as it works out quite a nice improvement imo)
- I've made the config option all-powerful, so if players don't want it in custom levels, they can control it. I think it does make sense, as you say, if they want the OG experience

walkawayy reviewed (approved):
Can't test atm but code looks good.

lahm86 committed:
Add showcase
